File: Python/Insert Interval.py
# Definition for an interval.
class Interval:
    def __init__(self, s=0, e=0):
        self.start = s
        self.end = e

class Solution:
    # @param intervals, a list of Intervals
    # @param newInterval, a Interval
    # @return a list of Interval
    def cmp1(self, x, y):
        if x.start < y.start:
            return -1
        elif x.start > y.start:
            return 1
        elif x.start == y.start:
            if x.end < y.end:
                return -1
            else:
                return 1

    def insert(self, intervals, newInterval):
        intervals.sort(self.cmp1)
        seq = []
        index = -1
        for i in range(len(intervals)):
            if intervals[i].start > newInterval.end:
                index = i
                break
            elif newInterval.start > intervals[i].end:
                seq.append(intervals[i])
            else:
                newInterval.start = min(newInterval.start, intervals[i].start)
                newInterval.end = max(newInterval.end, intervals[i].end)
        seq.append(newInterval)
        if index != -1:
            seq += intervals[index:]
        return seq

File: C++/Insert-Interval.cpp
/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */
class Solution {
public:
    vector<Interval> insert(vector<Interval>& intervals, Interval newInterval) {
        vector<Interval> res;
        int cur = 0;
        for (; cur < intervals.size(); cur ++) {
            if (intervals[cur].start > newInterval.end) {
                break;
            } else if (intervals[cur].end < newInterval.start) {
                res.push_back(intervals[cur]);
            } else {
                newInterval.start = min(newInterval.start, intervals[cur].start);
                newInterval.end = max(newInterval.end, intervals[cur].end);
            }
        }
        res.push_back(newInterval);
        for (; cur < intervals.size(); cur ++) {
            res.push_back(intervals[cur]);
        }
        return res;
    }
};